File: data/lib/remote/hash.py
"""Structured hash engine for EFA V2/V3 schema.

Primitives:
  ident_hash(uri_string)  → SHA-256 hex of identifier URI
  content_hash(bytes)     → SHA-256 hex of raw bytes

Structured hashes (v3 — only canonical JSON files are hashed):
  snapshot_hash(type, files)  → "efa:{type}:v3\\n" + metadata.json hash
  generation_hash(files)      → "efa:generation:v3\\n" + metadata.json hash
"""

from __future__ import annotations

import hashlib

from typing import Literal


SnapshotType = Literal["resource", "release"]
HASH_ALGORITHM = hashlib.sha256


def ident_hash(uri_string: str) -> str:
    """SHA-256 hex digest of an identifier URI string.

    Example:
        ident_hash("resource://tranquility/proto/ships.bin")
        → 64 lowercase hex chars
    """
    return HASH_ALGORITHM(uri_string.encode("utf-8")).hexdigest()


def content_hash(data: bytes) -> str:
    """SHA-256 hex digest of raw bytes (blob content)."""
    return HASH_ALGORITHM(data).hexdigest()


def _file_hash(filename: str, data: bytes) -> str:
    """Return 'filename <sha256_hex>' line for structured hash assembly."""
    return f"{filename} {HASH_ALGORITHM(data).hexdigest()}"


def snapshot_hash(snapshot_type: SnapshotType, files: dict[str, bytes]) -> str:
    """Compute structured snapshot hash.

    The hash input is built from only the canonical ``metadata.json`` file:

        "efa:{type}:v3\\n"
        "metadata.json <sha256>\\n"

    Args:
        snapshot_type: One of "resource", "release".
        files: Dict containing at least "metadata.json" → bytes.
    """
    if "metadata.json" not in files:
        raise ValueError("Missing required file: metadata.json")
    line = _file_hash("metadata.json", files["metadata.json"])
    payload = f"efa:{snapshot_type}:v3\n{line}\n"
    return HASH_ALGORITHM(payload.encode("utf-8")).hexdigest()


def generation_hash(files: dict[str, bytes]) -> str:
    """Compute structured generation hash.

    The hash input is built from the canonical ``metadata.json`` file only:

        "efa:generation:v3\\n"
        "metadata.json <sha256>\\n"
    """
    if "metadata.json" not in files:
        raise ValueError("Missing required file: metadata.json")
    line = _file_hash("metadata.json", files["metadata.json"])
    payload = f"efa:generation:v3\n{line}\n"
    return HASH_ALGORITHM(payload.encode("utf-8")).hexdigest()

File: lib/storage/repo/hash.dart
import "dart:convert";
import "dart:typed_data";

import "package:crypto/crypto.dart";

/// SHA-256 hex digest and structured hash utilities.
class RepoHash {
  const RepoHash._();

  /// Computes the SHA-256 hex digest of raw [bytes].
  static String hashBytes(Uint8List bytes) => sha256.convert(bytes).toString();

  /// Computes the SHA-256 hex digest of a UTF-8 encoded [string].
  static String hashString(String string) => hashBytes(utf8.encode(string));

  /// Computes the SHA-256 hex digest of raw content [bytes].
  ///
  /// Same as [hashBytes] — semantic alias for content hashing.
  static String hashContent(Uint8List bytes) => hashBytes(bytes);

  /// Computes the ident hash for a blob URI.
  ///
  /// ident_hash = SHA-256(uri_string) — used for blob storage addressing.
  static String hashIdent(String uri) => hashString(uri);

  // ── Structured hashes (agent/schemav2/schema.md §6.2, v3) ──────────────────

  /// Computes the structured resource snapshot hash (v3).
  ///
  ///   snapshot_hash = SHA-256(
  ///     "efa:resource:v3\n"
  ///     "metadata.json {sha256_of_canonical_metadata_json_bytes}\n"
  ///   )
  static String hashResourceSnapshot(String metadataJsonHash) =>
      hashString("efa:resource:v3\nmetadata.json $metadataJsonHash\n");

  /// Computes the structured release snapshot hash (v3).
  static String hashReleaseSnapshot(String metadataJsonHash) =>
      hashString("efa:release:v3\nmetadata.json $metadataJsonHash\n");

  /// Computes the structured generation hash (v3).
  ///
  ///   generation_hash = SHA-256(
  ///     "efa:generation:v3\n"
  ///     "metadata.json {hash}\n"
  ///   )
  static String hashGeneration({required String metadataJsonHash}) =>
      hashString("efa:generation:v3\nmetadata.json $metadataJsonHash\n");
}
